libraries/AP_HAL_ChibiOS/shared_dma.cpp

```diff
@@ -87,6 +87,7 @@ bool Shared_DMA::lock_stream(uint8_t stream_id)
         const thread_t* curr_owner = locks[stream_id].mutex.owner;
         chMtxLock(&locks[stream_id].mutex);
         cont = curr_owner != nullptr && curr_owner != locks[stream_id].mutex.owner;
+        osalDbgAssert(curr_owner != locks[stream_id].mutex.owner, "Recursive DMA lock");
```

Comment:
have you checked SPI sdcard? It may have recursive locks.
I'd rather have a check of the lock count. ie. locks[stream_id].mutex.cnt<10

Reply:
it might be worth having this as an internal error? that would work without --enable-asserts, and would show in a log or watchdog dump